File: june14/june19_Heaps2/selfheap.cpp
#include<iostream>
#include<vector>
using namespace std;

class heap{
    private:
        vector<int> list;
        bool type;//true for min
        void swap(int i,int j)
        {
            int ith=list[i];
            int jth=list[j];
            list[i]=jth;
            list[j]=ith;
        }
        bool ishp(int i,int j)
        {
            if(this->type==true)
                return list[i]<list[j];
            else
                return list[i]>list[j];
        }
        void upheapify(int i)
        {
            if(i==0)
                return;
            int pi=(i-1)/2;
            if(ishp(i,pi)){
                swap(i,pi);
                upheapify(pi);
            }
        }
        void downheapify(int i)
        {
            int li=2*i+1;
            int ri=2*i+2;
            int hpi=i;
            if(li<list.size() && ishp(li,hpi))
                hpi=li;
            if(ri<list.size() && ishp(ri,hpi))
                hpi=ri;

            if(hpi!=i)
            {
                swap(i,hpi);
                downheapify(hpi);
            }
        }
    public:
        heap(bool type)
        {
            this->type=type;
        }
        void push(int val)
        {
            list.push_back(val);
            upheapify(list.size()-1);
        }
        void pop()
        {
            swap(0,list.size()-1);
            list.pop_back();
            downheapify(0);
        }
        int top()
        {
            return list[0];
        }
        int size()
        {
            return list.size();
        }
};
int main(int argc,char** argv)
{
    heap hp(false);
    hp.push(10);
    hp.push(2);
    hp.push(50);
    hp.push(15);
    hp.push(70);

    while(hp.size()>0)
    {
        int val=hp.top(); hp.pop();
        cout<< val<<" ";
    }
}

File: june14/june19_Heaps2/selfheap.java
import java.util.*;
public class selfheap{
    private static ArrayList<Integer> list=new ArrayList<>();
    private static boolean type;//true for min - false for max
    selfheap(boolean type){
        this.type=type;
    }
    private static void swap(int i,int j){
        int ith=list.get(i);
        int jth=list.get(j);
        list.set(i,jth);
        list.set(j,ith);
    }
    private static boolean ishp(int i,int j){//if i has higher priority return true
        if(type==true){
            return list.get(i)<list.get(j);
        }else{
            return list.get(i)>list.get(j);
        }
    }
    private static void upheapify(int i){
        if(i==0)
            return;
        int pi=(i-1)/2;
        if(ishp(i,pi)==true){
            swap(i,pi);
            upheapify(pi);
        }
    }
    private static void downheapify(int i){
        int li=2*i+1;
        int ri=2*i+2;
        
        int hpi=i;//highest priority index
        if(li<list.size() && ishp(li,hpi)==true)
            hpi=li;
        if(ri<list.size() && ishp(ri,hpi)==true)
            hpi=ri;
        
        if(hpi!=i){
            swap(i,hpi);
            downheapify(hpi);
        }
    }
    public static void add(int val){
        list.add(val);
        upheapify(list.size()-1);
    }
    public static int peek(){
        return list.get(0);
    }
    public static int remove(){
        int rv=peek();
        swap(0,list.size()-1);
        list.remove(list.size()-1);
        downheapify(0);
        return rv;
    }
    public static int size(){
        return list.size();
    }
    public static void main(String[] args){
        selfheap pq=new selfheap(false);
        pq.add(10);
        pq.add(5);
        pq.add(20);
        pq.add(2);
        pq.add(60);

        while(pq.size()>0){
            int val=pq.remove();
            System.out.print(val+" ");
        }
    }
}